text/0000-layer-metadata-files.md

```diff
@@ -0,0 +1,67 @@
+# Meta
+[meta]: #meta
+- Name: Disambiguate layer metadata files from Application metadata files
+- Start Date: 2021-03-11
+- Author(s): [samj1912](https://github.com/samj1912)
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: (put "N/A" unless this replaces an existing RFC, then link to that RFC)
+
+# Summary
+[summary]: #summary
+
+This is a proposal to change the location of layer specific TOML metadata files from application specific TOML metadata files like `launch.toml` or `build.toml` in order to disambiguate special metadata files from layer metadata files that can have similar names.
+
+# Motivation
+[motivation]: #motivation
+
+Currently all the metadata files live under the `<layers>` folder. This means that application metadata files like `launch.toml`, `build.toml` or `store.toml` share the same directory as layer metadata files. Since the layers could possibly be named `launch`, `build` or `store` this leads to ambiguity, conflicts and/or restrictions on layer names and their associated metadata files that should not exist. This also prevents us from possibly adding new metadata files in the future without breaking backwards compatibility with buildpacks who may have been creating layers with the same name. This proposal seeks to disambiguate special metadata files from free-form layer metadata files.
+
+# What it is
+[what-it-is]: #what-it-is
+
+The proposal is as follows - 
+
+All layer metadata files should be prefixed with the string `layer.`. For example, the current implementation expects layer specific metadata file for a layer `<layers>/example` to be located at `<layers>/example.toml`, we now propose that the associated layer metadata file be located at `<layers>/layer.example.toml`.
+
+This RFC would also prohibit buildpack authors from putting any files in the `<layers>` directory that is not currently a recognized metadata file or begins with the `layer.` prefix. This is to allow the buildpack API to accommodate more of special metadata files like `build.toml` without breaking any backwards compatibility in an unambiguous fashion.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+The lifecycle would look for layer metadata files at `<layers>/layer.<layer>.toml` instead of `<layers>/<layer>.toml`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Why should we *not* do this?
+
+This will break backwards compatibility for numerous buildpacks that don't use a set of bindings like `libcnb`.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- What other designs have been considered?
+
+We could have separate folders for layers and their metadata files and the application metadata files.
+
+- Why is this proposal the best?
+
+It seemed easier to add a prefix instead of creating a sub-folder in order to provide proper namespacing for layer metadata.
```

Comment:
It might be only me but although it seemed easier to add a prefix, I think that it'll be clearer to create a sub-folder.
I think about 2 examples where adding adding a prefix could be a little weird:
* A long layer name with dots in the middle.
For example: `foo.bar.test` will become `layer.foo.bar.test`.
* A layer name that already contains the word "layer".
For example:`layer.test` will become `layer.layer.test`, and `my.test.layer` will become `layer.my.test.layer`.

I don't have a strong opinion on this, just a slight preference.

Reply:
I agree with this sentiment. 

I would like to throw another potential alternative. Instead of making the layer metadata file _more_ special, we make the other files (`build.toml`, `launch.toml` `store.toml`) special by prefixing those.

For example:

Prefixing non-layer metadata files with `cnb.` and adding restrictions to the spec such as "Layers may not be prefixed with `cnb.`".

This would yield a file structure such as...

```
.
└── <layer>/
    ├── cnb.build.toml
    ├── cnb.launch.toml
    ├── cnb.store.toml
    ├── <layer>.toml
    └── */
        └── <other-content>
```

We do something similar with [environment variables](https://github.com/buildpacks/spec/blob/main/buildpack.md#provided-by-the-platform):

> The environment variable prefix CNB_ is reserved. It MUST NOT be used for environment variables that are not defined in this specification or approved extensions.

Reply:
@jkutner also proposed a similar idea during the WG meeting. This might be a good compromise between our current system and the primary proposal! Will add it as an alternative tomorrow :)

Reply:
Added this alternative as well.